source/FreeRTOS_IPv6.c

```diff
@@ -270,34 +262,60 @@ static void xGetIPv6MulticastGroupID( const IPv6_Address_t * pxIPv6Address,
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Check if the IP-address is an IPv6 loopback address.
+ *
+ * @param[in] pxAddress The IP-address being checked.
+ *
+ * @return pdTRUE if the IP-address is a loopback address or else, pdFALSE.
+ */
+BaseType_t xIsIPv6Loopback( const IPv6_Address_t * pxAddress )
+{
+    BaseType_t xReturn = pdFALSE;
+
+    if( memcmp( pxAddress->ucBytes, FreeRTOS_in6addr_loopback.ucBytes, ipSIZE_OF_IPv6_ADDRESS ) == 0 )
+    {
+        xReturn = pdTRUE;
+    }
+
+    return xReturn;
+}
+
 #if ( ipconfigETHERNET_DRIVER_FILTERS_PACKETS == 0 )
 
 /**
- * @brief Check if the packet is a loopback packet.
+ * @brief Check if the packet is an illegal loopback packet.
  *
- * @param[in] pxIPv6Header The IP packet in pxNetworkBuffer.
+ * @param[in] pxIPv6Header The IP-header of the packet.
  *
- * @return Returns pdTRUE if it's a legal loopback packet, pdFALSE if not .
+ * @return Returns pdTRUE if the packet should be stopped, because either the source
+ *         or the target address is a loopback address.
  */
 /* MISRA Ref 8.9.1 [File scoped variables] */
 /* More details at: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/MISRA.md#rule-89 */
 /* coverity[misra_c_2012_rule_8_9_violation] */
 /* coverity[single_use] */
-    static BaseType_t xIsIPv6Loopback( const IPHeader_IPv6_t * const pxIPv6Header )
+    BaseType_t xBadIPv6Loopback( const IPHeader_IPv6_t * const pxIPv6Header )
     {
         BaseType_t xReturn = pdFALSE;
         const NetworkEndPoint_t * pxEndPoint = FreeRTOS_FindEndPointOnIP_IPv6( &( pxIPv6Header->xSourceAddress ) );
 
         /* Allow loopback packets from this node itself only. */
-        if( ( pxEndPoint != NULL ) &&
-            ( memcmp( pxIPv6Header->xDestinationAddress.ucBytes, FreeRTOS_in6addr_loopback.ucBytes, sizeof( IPv6_Address_t ) ) == 0 ) &&
-            ( memcmp( pxIPv6Header->xSourceAddress.ucBytes, pxEndPoint->ipv6_settings.xIPAddress.ucBytes, sizeof( IPv6_Address_t ) ) == 0 ) )
+        if( pxEndPoint != NULL )
         {
-            xReturn = pdTRUE;
+            BaseType_t x1 = ( xIsIPv6Loopback( &( pxIPv6Header->xDestinationAddress ) ) != 0 ) ? pdTRUE : pdFALSE;
+            BaseType_t x2 = ( xIsIPv6Loopback( &( pxIPv6Header->xSourceAddress ) ) != 0 ) ? pdTRUE : pdFALSE;
+
+            if( x1 != x2 )
```

Comment:
This logic is not very clear.Isn't it enough to check only for destination address.
 In my understanding, We should only process packets for which the destination address is loopback. Source address can only be loopback if there is a bridging/routing supported in the device and as of now our stack does not support bridging/routing. And from an external device only the destination address can be loopback. Please suggest if I am missing anything here.

Reply:
I already answered to this question here above.